[ShapeOpt] feature model controller update

armingeiser opened the pull request:
The settings for the shape optimization have been adapted:

* The model is used in the application to make it ready for #2417 
* Now there is a "model_settings" block that contains:
    * model import 
    * mesh motion
    * damping
  The two latter are moved from the "design_variables" block
* Mesh motion settings block is now identical to the "default" settings of the mesh motion analysis (could potentially be excluded from the main settings file as we did for the primal settings - But in a separate PR) 

Tests and examples are adapted and running.

closes #2925 
closes #2979

armingeiser committed:
update ShapeOpt to use the model and update tests and test examples

armingeiser committed:
add new model settings and clean code

armingeiser committed:
cleaning

armingeiser committed:
adapt tests to model_settings

armingeiser committed:
add .csv of tests to gitignore

armingeiser committed:
update test examples to model_settings

armingeiser committed:
remove deprecated argument

dbaumgaertner commented on applications/ShapeOptimizationApplication/python_scripts/mesh_controller_with_solver.py:
```diff
@@ -48,21 +50,26 @@ def __init__(self, MeshSolverSettings, OptimizationModelPart):
-
-        self.MeshSolverSettings["problem_data"].AddEmptyValue("model_part_name")
-        self.MeshSolverSettings["problem_data"]["model_part_name"].SetString(OptimizationModelPart.Name)
+        # add default problem data parameters for mesh motion analysis
```
Looks much better.

dbaumgaertner commented on applications/ShapeOptimizationApplication/test_examples/01_Strain_Energy_Minimization_3D_Hook/optimization_parameters.json:
```diff
@@ -1,18 +1,81 @@
 {
     "optimization_settings" : {
+        "model_settings" : {
```
Looks way better than before!

dbaumgaertner commented on applications/ShapeOptimizationApplication/tests/algorithm_penalized_projection_test/run_test.py:
```diff
@@ -35,14 +34,14 @@ def __init__( self ):
     # --------------------------------------------------------------------------
-    def __CalculateValue( self ):
-        constrained_node = optimization_model_part.GetNodes()[self.constrained_node_id]
+    def __CalculateValue( self, current_design ):
```
That is, how it was originally supposed to be used. I like that change.

dbaumgaertner commented on applications/ShapeOptimizationApplication/python_scripts/optimizer_factory.py:
```diff
@@ -24,9 +24,25 @@
+def CreateOptimizer(optimization_settings, model, external_analyzer=EmptyAnalyzer()):
 
-    model_part_controller = model_part_controller_factory.CreateController(optimization_settings, optimization_mdpa)
+    default_settings = Parameters("""
```
Good to have that now.

dbaumgaertner commented on applications/ShapeOptimizationApplication/test_examples/.gitignore:
```diff
@@ -0,0 +1 @@
+**/*.csv
```
Is this change in the gitignore done intentionally?

armingeiser replied:
yes, because the results of the test examples where always shown as "untracked".

dbaumgaertner committed:
minor cosmetics 

changed name of parameter to be consistent and also changed at one point the order of functions.

armingeiser committed:
Merge remote-tracking branch 'origin/master' into ShapeOpt/feature-model-controller-update

dbaumgaertner reviewed (dismissed):
Just a few comments prior to the following commits.

armingeiser committed:
simplify extraction of surface nodes

armingeiser committed:
add process to assign variable from other variable

armingeiser committed:
use new process by default in mesh controller with solver

armingeiser commented:
@dbaumgaertner the new process is there and works nicely. You can review it again

armingeiser committed:
simplify process after discussion

philbucher commented on applications/ShapeOptimizationApplication/python_scripts/fix_vector_variable_process.py:
```diff
@@ -0,0 +1,82 @@
+        raise Exception("Expected input shall be a Parameters object, encapsulating a json string")
+    return FixVectorVariableProcess(model, settings["Parameters"])
+
+class FixVectorVariableProcess(KratosMultiphysics.Process):
```
it seems like this process does the same as the one in the core
didn't you tell me that you want to make sth dependent on a variable?

armingeiser replied:
That was the original idea. But now it is only fixing the variable (The process in the core also sets the values to a value  defined in the json). I just want to preserve the values that are already on the nodes.

philbucher replied:
Ok I see
Then we should discuss if we can move it to the core, I also would need this functiionality :)

armingeiser replied:
Should we merge it first here and then move it?

philbucher replied:
I think you could put it in the core right away
For this I would also split it into scalar and vector, same as for the `assign_`-processes in the core

philbucher commented on applications/ShapeOptimizationApplication/custom_utilities/geometry_utilities.h:
```diff
@@ -128,18 +128,13 @@ class GeometryUtilities
 
     // --------------------------------------------------------------------------
-    void ExtractBoundaryNodes( std::string const& NewSubModelPartName )
+    void ExtractBoundaryNodes( std::string const& BoundarySubModelPartName )
```
`rBoundarySubModelPartName` ;)

armingeiser committed:
code style and codacy

armingeiser committed:
switch to core process

armingeiser committed:
Merge branch 'master' into ShapeOpt/feature-model-controller-update

armingeiser committed:
move projection of normals to model part controller

armingeiser committed:
Merge branch 'master' into ShapeOpt/feature-model-controller-update

armingeiser committed:
delete duplicated process after it was moved to core

armingeiser committed:
delete unnecessary filename if use_input_model_part

armingeiser commented:
@dbaumgaertner This is ready for a final review now.

dbaumgaertner commented on applications/ShapeOptimizationApplication/test_examples/01_Strain_Energy_Minimization_3D_Hook/optimization_parameters.json:
```diff
@@ -1,18 +1,80 @@
             "use_kratos" : true,
             "kratos_response_settings":{
                 "response_type"          : "mass",
+                "model_part_name"    : "3D_Hook",
```
Why do we have to specify this model_part_name and model_import_settings, can't we use default values?

armingeiser replied:
Like for other responses (e.g. Strain energy), the name and also the import mode of the model part has to be given in the settings. For the mechanichal responses, this information is in the `primal_settings.json`, for the mass response function, we don't have an external settings file, thus giving the information directly here.

dbaumgaertner committed:
minor cosmetic

dbaumgaertner committed:
improved readability

dbaumgaertner committed:
adjusted naming of damping function

to better reflect the actual damping settings

dbaumgaertner commented:
@armingeiser Just some suggestions for minor improvements. In case these are fine, you can go on and merge. The new setup improves a lot the readability and the overall code layout. Very good job! :1st_place_medal:
